### DISCUSSION TITLE
API Calls Not Working Supabase (Testing on Android Emulator and APK)

### DISCUSSION THREAD

**Discussion by Matignaciom** [supabase/supabase#29008] [Questions] [answered]
I'm facing an issue in my application where API calls are not working for all users in my auth.users table, except for a specific user with ID `92bef010-d430-4eb6-b449-8cab4705c12a`.

**The rest of the users are unable to access the API calls**. This issue is occurring with my app that is being tested via an APK installed on my Android device and through the Android Studio emulator. The app is not yet published on the Play Store or App Store. I am using Supabase and have verified that the user data is configured correctly. How can I investigate and resolve this issue? What steps should I take to ensure that all users have the proper access to the API calls?

**Is this a Supabase bug or is something else happening? Because my first user (the one who was created first) can see all the API calls, but with other users, nothing works. How can I check this with SQL or something else? Any ideas? I don't know what to do anymore... I’m not sure if it’s a Supabase issue or what... urgent help needed.**

**Comment by GaryAustin1** [comment, marked as answer]
I doubt it is a Supabase issue as you are the only one reporting it.

It is not clear what you mean by "can see all the API calls".

Do you have some function somewhere that checks auth.uid() for a hard coded value for RLS maybe?  Once again not knowing what "can see" and API call means.

I see you have a Discord message open also.  I'll follow up there.

**Reply by Matignaciom** [reply]
@GaryAustin1 The problem I'm having is that in my App the first user I created can see the call api perfectly but the others cannot, it is inserted and everything correctly but I don't understand why the other users can't see the results of the call apis in my project.Is there any way to check if it's because a specific user doesn't have some permission or something like that that is needed to be able to see the results of the call apis? Because I currently can't see the result of the call APIs with my other users.


Here's how I see it from another user:

![Screenshot_2024-08-30-17-27-44-203_com.acsch.comunidadacsch.jpg](https://github.com/user-attachments/assets/ed4cbea0-e62d-4a7d-95c7-dc9e873cda44)

And here's how I see it with my first user created:

![IMG_20240830_172816.jpg](https://github.com/user-attachments/assets/e36b9236-08a7-415d-8ebc-1a8bc3b6bec3)


The logic and all the operation is correctly configured but for some reason with one profile I can see the result of my call api and with another I can't, it's strange..

**Reply by GaryAustin1** [reply]
I'm not sure anyone will be able to help you with the information you present.  Pictures of your app are not helpful.
Errors or specific calls (with no data?) from Supabase might be.  What is the RLS applied to those operations?  What is different between the first user and the others?  How did the users signup?

As I said also this is usually easier on Discord because of back and forth up to you.
But you are starting to head into the weekend now also.

**Reply by Matignaciom** [reply]
@GaryAustin1 The two users have the same type of profile (Maestro Titular) and logically everything works well, but for some strange reason, my newly registered users cannot get a response from the API calls. These API calls fail for all users created except for the first one I created from the beginning.

Here is the RLS of my users table:
![image](https://github.com/user-attachments/assets/61121ea9-ed4b-45cb-97fd-2458445de169)

**Here's how I insert a new teacher**:

```
CREATE OR REPLACE FUNCTION public.insertar_maestro(p_email character varying, p_nombre character varying, p_direccion character varying, p_telefono character varying, p_fecha_nacimiento text, p_tipo_perfil character varying, p_contrasena character varying, p_iglesia_id uuid) RETURNS uuid AS $$CREATE OR REPLACE FUNCTION public.insertar_maestro(p_email character varying, p_nombre character varying, p_direccion character varying, p_telefono character varying, p_fecha_nacimiento text, p_tipo_perfil character varying, p_contrasena character varying, p_iglesia_id uuid)
 RETURNS uuid
 LANGUAGE plpgsql
 SECURITY DEFINER
AS $function$
DECLARE
    v_usuario_id UUID;
    v_fecha_nacimiento DATE;
BEGIN
    -- Convertir la fecha de nacimiento de TEXT a DATE usando el formato DD/MM/YYYY
    v_fecha_nacimiento := TO_DATE(p_fecha_nacimiento, 'DD/MM/YYYY');
    
    -- Insertar el nuevo maestro en la tabla usuarios
    INSERT INTO usuarios (
        id, email, nombre, direccion, telefono, fecha_nacimiento, tipo_perfil, contrasena, fecha_creacion, habilitada, iglesia_id
    )
    VALUES (
        gen_random_uuid(), 
        p_email, 
        p_nombre, 
        p_direccion, 
        p_telefono, 
        v_fecha_nacimiento, 
        p_tipo_perfil::tipo_perfil, -- Conversión explícita al tipo personalizado
        p_contrasena, 
        NOW(), 
        TRUE, 
        p_iglesia_id
    )
    RETURNING id INTO v_usuario_id;
    
    -- Devolver el ID del nuevo usuario creado
    RETURN v_usuario_id;

EXCEPTION
    WHEN others THEN
        -- Manejo de errores genérico
        RAISE EXCEPTION 'Error al insertar el maestro: %', SQLERRM;
        RETURN NULL;
END;
$function$
$$ LANGUAGE plpgsql;
```

**Reply by GaryAustin1** [reply]
You still have not shown your API call to Supabase or RLS on any data table they might be accessing.  
Are you even talking about API calls to Supabase or to your own server?

You are showing a profile table with basically open RLS (anyone can access it), but why is the profile table important?  Is that used in RLS somewhere else?

**Reply by Matignaciom** [reply]
I make an API call to Supabase here, and the result is positive, but in the app, I don't receive a response from any user other than the one I mentioned earlier. For now, I have the RLS table open for testing.

![image](https://github.com/user-attachments/assets/b918f586-a303-4d09-a885-5559a6226102)

**Reply by Matignaciom** [reply]
Hello, I was able to solve it! It was a problem of logic in my function, I was able to correct it, thank you!

**Reply by fullupe** [reply]
can you share that logic ....because i have similar issue everything work fine in emmulator but in apk i cant insert but i ca read data....RLS is blocked

**Reply by Matignaciom** [reply]
> can you share that logic ....because i have similar issue everything work fine in emmulator but in apk i cant insert but i ca read data....RLS is blocked

I defined `SECURITY DEFINER` in my function, which enabled me to perform certain operations with elevated privileges. Here's an improved example:

```sql
CREATE OR REPLACE FUNCTION my_secure_function()
RETURNS VOID
LANGUAGE sql
SECURITY DEFINER
AS $$
    INSERT INTO sensitive_table (column1, column2)
    VALUES ('value1', 'value2');
$$;
```

Explanation:  
**SECURITY DEFINER:** Allows the function to execute with the permissions of the user who defined the function (not the one who executes it).  
Make sure to carefully limit the permissions of the user defining the function to avoid potential security risks.

**Comment by henrybrown9907-cmyk** [comment]
This is most likely not a Supabase/APK issue. If one user works while other authenticated users fail, check your authentication logic and RLS policies first.

In particular, look for any function or RLS policy that uses `auth.uid()` and accidentally compares it against a hard-coded user ID[.](https://apkjunction.com.pk/) Also verify that the API request is actually being made with the authenticated user's session/JWT.

For debugging:

1. Check the RLS policies on the table being accessed.
2. Search your SQL functions and policies for `auth.uid()`.
3. Make sure you aren't using a hard-coded UUID anywhere in the authorization logic.
4. Test the same request with different authenticated users and compare the returned errors/status codes.
5. If an operation genuinely needs elevated privileges, a carefully designed `SECURITY DEFINER` function can be used, but it should be restricted and written securely.

In the original discussion, the issue ultimately turned out to be a logic problem in the user's function rather than a Supabase bug.

If your problem is specifically INSERT works in the emulator but fails in the Apk,, also check whether the APK is using the same Supabase URL, anon key, authenticated session, and RLS context as the emulator.

**Comment by theodoreashford7-cell** [comment]
I’d first check your Supabase RLS policies rather than assuming it’s a Supabase bug. If one user works but newly created users don’t, compare their JWT claims/roles and test the affected tables with RLS enabled. Also check the API response status and Supabase logs to see whether requests are being rejected by a policy[.](https://minecraftmod.org/) A simple SQL query can help verify whether the policies allow the expected users to a

**Comment by henrybrown9907-cmyk** [comment]
I’d compare the working and failing users’ JWTs, RLS policies, and `auth.uid()` results first, since that can quickly show whether the issue is authentication or database permissions. The same kind of user[-](https://joiiplay.com/)specific testing is useful when troubleshooting an Android game across different devices and environments.
